sdk/core/Azure.Core/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -47,59 +72,150 @@ public bool TryGetValue(ulong key, out object? value)
+                    }
+
+                    if (IsSecond(key))
+                    {
+                        _second = default;
+                        _count--;
+                        return true;
+                    }
+
+                    return false;
+                default:
+                    if (IsFirst(key))
+                    {
+                        _first = _second;
+                        _second = _rest[0];
+                        _count--;
+                        Array.Copy(_rest, 1, _rest, 0, _count - 2);
+                        _rest[_count - 2] = default;
+                        return true;
+                    }
+
+                    if (IsSecond(key))
+                    {
+                        _second = _rest[0];
+                        _count--;
+                        Array.Copy(_rest, 1, _rest, 0, _count - 2);
+                        _rest[_count - 2] = default;
+                        return true;
+                    }
+
+                    for (var i = 0; i < _count - 2; i++)
+                    {
+                        if (IsRest(i, key))
+                        {
+                            _count--;
+                            Array.Copy(_rest, i + 1, _rest, i, _count - 2 - i);
+                            _rest[_count - 2] = default;
+                            return true;
+                        }
+                    }
+
+                    return false;
             }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool IsFirst(TKey key) => _first.Key.Equals(key);
 
-            if (_rest.Length <= _count)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool IsSecond(TKey key) => _second.Key.Equals(key);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool IsRest(int index, TKey key) => _rest[index].Key.Equals(key);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void AddInternal(TKey key, TValue value)
+        {
+            switch (_count)
             {
-                var larger = ArrayPool<KVP>.Shared.Rent(_rest.Length << 1);
-                _rest.CopyTo(larger, 0);
-                var old = _rest;
-                _rest = larger;
-                ArrayPool<KVP>.Shared.Return(old, true);
+                case 0:
+                    _first = new Kvp(key, value);
+                    _count = 1;
+                    return;
+                case 1:
+                    if (IsFirst(key))
+                    {
+                        _first = new Kvp(_first.Key, value);
+                    }
+                    else
+                    {
+                        _second = new Kvp(key, value);
+                        _count = 2;
+                    }
+
+                    return;
+                default:
+                    if (_rest == null)
+                    {
+                        _rest = ArrayPool<Kvp>.Shared.Rent(8);
```

Comment:
Do we ever return this? Should this type implement IDisposable?

Reply:
We do, and there is a `Dispose` method that is called in `PipelineRequest.Dispose`. @christothes , is there a reason why we don't implement `IDisposable` here?

Reply:
I don't recall now exactly, but when I added the interface, it caused some problem.

Reply:
Ok, let's skip the interface then

Reply:
Was the issue that we needed to make another type IDisposable? 😆